### PR TITLE
[Gohai][ASC-572] implement new api for cpu package

### PR DESCRIPTION
**Pull request by pgimalac**
### What does this PR do?

Implement the new API for the Cpu package of Gohai, following what was described in the [Gohai design RFC](https://github.com/DataDog/architecture/pull/1064). 

### Motivation

Have a cleaner API for the Cpu package, which also reduce the risk of bugs when using the API and make sure the values have a standard format.

### Additional Notes

The data collected on Linux should be exactly the same.

On Darwin and Windows, we used to return cpu frequency as an int but now return it as a float (to align with the behavior on Linux).

The other change introduced in this PR is to not return errored fields in the JSON on Windows (the default value of the type was returned before).

I tested and got the exact same output on the following Os/Arch:
- Ubuntu amd64 and arm64
- AmazonLinux amd64 and arm64
- Debian amd64 and arm64
- Redhat amd64 and arm64
- Suse amd64 and arm64
- Windows amd64
- Darwin arm64

### Describe how to test/QA your changes

Check that the Cpu package still works with correct values on Linux, Darwin and Windows. 

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by hestonhoffman — APPROVED** [review]
Left some edits for the release notes

File: pkg/gohai/cpu/cpu_darwin.go
Comment (ogaca-dd) [inline pkg/gohai/cpu/cpu_darwin.go:15]: Did you consider having `getSysctl[T](call func(string) (T, error), key string) T` so the caller doesn't have to define a `cast` function (The caller can do the conversion without a callback)?

File: pkg/gohai/cpu/cpu_linux_arm64.go
Comment (ogaca-dd) [inline pkg/gohai/cpu/cpu_linux_arm64.go:154]: Why not setting the error for `cpuInfo.fillProcCPUInfo`?

File: pkg/gohai/cpu/cpu_linux_arm64.go
Comment (ogaca-dd) [inline pkg/gohai/cpu/cpu_linux_arm64.go:170]: What do you think about having a more generic error as I guess `os.ReadDir` can return an error also when the permissions are not enough to read the folder?

File: pkg/metadata/inventories/host_metadata.go
Comment (ogaca-dd) [inline pkg/metadata/inventories/host_metadata.go:102]: What do you think about having a `ValueOrDefault` method as relying on the value when there is an error is uncommon.

**Review by ogaca-dd — APPROVED** [review]
LGTM.
Please note that I didn't check that the code behaves the same as before as it is not really possible to do during a review.

File: pkg/collector/corechecks/system/cpu/cpu_windows.go
Comment (pgimalac) [reply pkg/collector/corechecks/system/cpu/cpu_windows.go:152]: `GetCpuInfo` never returned a non-`nil` error, in case of error this code would just use the default value, but as it really seemed like the intent was to return in case of error I kept it that way.

I can change that behavior if @DataDog/windows-agent wants to just ignore the error.

File: pkg/gohai/cpu/cpu_linux_arm64.go
Comment (pgimalac) [reply pkg/gohai/cpu/cpu_linux_arm64.go]: Very good remark, changed to `could not read /sys/devices/system/node` in 0ab3c133b66a7e383514c72e4cce3618df112ae7

File: pkg/metadata/inventories/host_metadata.go
Comment (pgimalac) [reply pkg/metadata/inventories/host_metadata.go:102]: Good idea, I'll implement that in a follow-up PR as it also changes code in other gohai packages

File: pkg/gohai/cpu/cpu_linux_arm64.go
Comment (pgimalac) [reply pkg/gohai/cpu/cpu_linux_arm64.go:154]: Moved the error handling logic directly inside `fillProcCPUInfo` 👍

File: pkg/gohai/cpu/cpu_windows.go
Comment (iglendd) [inline pkg/gohai/cpu/cpu_windows.go:233]: Previously it was int and now is float. I check registry it is DWORD

File: pkg/collector/corechecks/system/cpu/cpu_windows.go
Comment (pgimalac) [reply pkg/collector/corechecks/system/cpu/cpu_windows.go:156]: I wondered too, my best guess is that it was used to compute the number of cpu cycles, and it avoided converting it to a float when doing the division
It's been like that since the creation of the file in https://github.com/DataDog/datadog-agent/pull/1725

File: pkg/gohai/cpu/cpu_windows.go
Comment (pgimalac) [reply pkg/gohai/cpu/cpu_windows.go:233]: This is not an error: on Linux the CPU frequency is collected as a float (and it is sent as such in the metadata), one of the goals of the new API is that each piece of information has the same format/type no matter the platform (it helps preventing parsing issues for example), having it as a float on Windows too makes the API more uniform (and I think since it's a frequency in Mhz it makes sense to have a float)

File: pkg/gohai/cpu/cpu_darwin.go
Comment (ogaca-dd) [reply pkg/gohai/cpu/cpu_darwin.go]: What do you think about making `getSysctlString` returning `Value[string], error` instead of `string, error`?

**Review by ogaca-dd — APPROVED** [review]

File: pkg/gohai/cpu/cpu.go
Comment (pgimalac) [reply pkg/gohai/cpu/cpu.go:43]: On Linux AMD64 we have the total cache size but not the size of each level, the reason is just that no one implemented it, but I could do it if you have a use for it
Note that these values aren't even displayed in the "system information" page in the app

**Review by amenasria — APPROVED** [review]

**Review by amenasria — APPROVED** [review]
LGTM as long as the tests pass (once you merged / rebase `main` into your branch to fix the cython issue)

**Review by iglendd — APPROVED** [review]

File: pkg/process/checks/system_info_windows.go
Comment (robertjli) [inline pkg/process/checks/system_info_windows.go:37]: I'm curious, why do we check for an error for `cpuInfo.CPUPkgs.Value()` but not `cpuInfo.CPULogicalProcessors.Value()`?

**Review by robertjli — APPROVED** [review]
LGTM, just one question

File: pkg/process/checks/system_info_windows.go
Comment (pgimalac) [reply pkg/process/checks/system_info_windows.go:37]: I kept the previous behavior which is to ignore errors except specifically for the `CPUPkgs` field, because there is a division by that value.
The previous implementation assumed that `cpu.GetCpuInfo()` would return an error if any error occurs (in particular when collecting `CPUPkgs`) but the map would actually contain the default value if it failed to collect the field.
I replaced that error checking with the one of `CpuPkgs`.

Feel free to tell me if you think that function should return an error if any of the field fails to be collected.

Comment (robertjli) [reply pkg/process/checks/system_info_windows.go:37]: I think that makes sense, thank you!

**Pull request merged**
